OAS Guide: Read/Write-only in Schema (#296)
bump-sh/docs: src/_guides/openapi/specification/v3.1/data-models/schema-and-data-types.md

  id:
    type: string
    readOnly: true
  username:
    type: string
  date_of_birth:
    type: string
    format: date-time
  password:
    type: string
    writeOnly: true
  created_at:
    type: string
    format: date-time
    readOnly: true
```

By using these, a single schema can serve both requests and responses. For example, a `User` schema can send `password` during creation (`POST`), but exclude it in the `GET` response, while fields like `id` and `created_at` are only returned. This approach reduces duplication, making schemas easier to maintain.

These two keywords are considered "annotations" in JSON Schema, which means they are only there for various bits of tooling to do something with if they like. There is no requirement for tools to do anything in particular, but a common convention for most documentation tools will be to skip listing a `readOnly` property for a HTTP POST/PATCH request body, and similarly slip documenting `writeOnly` properties in a response body. 

This same approach extends to example request/responses of JSON data generated by documentation tools, and also mock servers, and even change the parameters available for generated SDK code.

> There are some complexities here with HTTP PUT because you're meant to be sending the entire resource each time whether something is writeable or not, but if a property was removed from a sample request for a HTTP PUT request that mean: a) "send it or not, we don't care", or b) not sending it will result in it being removed. Seeing as tooling varies on this, and API implementations vary in how they interpret missing values in PUT, you just need to check the tools you use do what your API expects.

Reviewer: ```suggestion
> There are some complexities here with HTTP PUT because you're meant to be sending the entire resource each time whether something is writeable or not, but if a property was removed from a sample request for an HTTP PUT request, that means: a) "send it or not, we don't care", or b) not sending it will result in it being removed. Seeing as tooling varies on this, and API implementations vary in how they interpret missing values in PUT, you just need to check the tools you use do what your API expects.
{: .info }
```

For more information on JSON Schema validation keywords, you can refer to the [JSON Schema Validation documentation](https://json-schema.org/learn/validation.html).

## readOnly & writeOnly

> JSON Schema provides `readOnly` and `writeOnly` boolean keywords, which are really helpful in the context of an API, because resources usually available in two flavours: the representation of a resource in a request body, and the representation of the resource in a response body.

Reviewer 2: ```suggestion
JSON Schema provides `readOnly` and `writeOnly` boolean keywords, which are really helpful in the context of an API, because resources are usually available in two flavours: the representation of a resource in a request body, and the representation of the resource in a response body.
```

- `readOnly: true` indicates that a value should not or cannot be be modified, but can be seen (e.g., `id`, `created_at`).

- `writeOnly: true` indicates that a value may be set, but will remain hidden (e.g., `password`, or PII like the `cvc` security code on a credit card).

```yaml
type: object
properties:
  id:
    type: string
    readOnly: true
  username:
    type: string
  date_of_birth:
    type: string
    format: date-time
  password:
    type: string
    writeOnly: true
  created_at:
    type: string
    format: date-time
    readOnly: true
```

By using these, a single schema can serve both requests and responses. For example, a `User` schema can send `password` during creation (`POST`), but exclude it in the `GET` response, while fields like `id` and `created_at` are only returned. This approach reduces duplication, making schemas easier to maintain.

> These two keywords are considered "annotations" in JSON Schema, which means they are only there for various bits of tooling to do something with if they like. There is no requirement for tools to do anything in particular, but a common convention for most documentation tools will be to skip listing a `readOnly` property for a HTTP POST/PATCH request body, and similarly slip documenting `writeOnly` properties in a response body. 

Reviewer 2: ```suggestion
These two keywords are considered “annotations” in JSON Schema, which means they are only there for various bits of tooling to do something with if they like. There is no requirement for tools to do anything in particular, but a common convention for most documentation tools will be to skip listing a `readOnly` property for a HTTP POST/PATCH request body, and similarly skip documenting `writeOnly` properties in a response body. 
```

This same approach extends to example request/responses of JSON data generated by documentation tools, and also mock servers, and even change the parameters available for generated SDK code.

> > There are some complexities here with HTTP PUT because you're meant to be sending the entire resource each time whether something is writeable or not, but if a property was removed from a sample request for an HTTP PUT request, that means: a) "send it or not, we don't care", or b) not sending it will result in it being removed. Seeing as tooling varies on this, and API implementations vary in how they interpret missing values in PUT, you just need to check the tools you use do what your API expects.

Reviewer 2: :ok_hand: puuur-fect thanks for the precision on this. BTW it makes a lot of sense not to hide `readOnly` properties from `PUT` request bodies, maybe we should implement this in Bump? :thinking:

Author: It depends how people are implementing PUT. Rails APIs have been using PUT and PATCH as an alias for eachother for decades and others have followed that convention...
